### ISSUE TITLE
Allow selection of an attribute to be the Edge ID when writing GraphML

### ISSUE THREAD

**Issue by dschult** [networkx/networkx#4353]
See #4347 and also #3960 

It looks like we should make it possible to select an edge attribute to be the "edge id" when writing GraphML.

**Comment by sestys** [comment]
Hey, 
It seems no one is working on this so I will do it. I just have few questions to start:
should any attribute be allowed to be used as the "edge id"? Also, should we then check for id uniqueness for parallel edges in MultiGraph?

**Comment by dschult** [comment]
I'm open to other opinions, but it seems to me we should let any edge attribute be allowed to be used as the edge id.  And my preference is to enforce id uniqueness within the node-pair that has multiedges. But not to enforce uniqueness across the whole network.  

Perhaps we can put an example into the docstring for how to change the edge keys to be unique (see #4347) across the whole network.

Thanks!

**Comment by berlincho** [comment]
Hi,

I am interested in this issue and #4347. I have read the [document](http://graphml.graphdrawing.org/primer/graphml-primer.html#GraphEdge) of graphml that the edge `id` is an optional field. 

> 2.3.3 Declaring an Edge
> Optionally an identifier for the edge can be specified with the XML Attribute id. When it is necessary to reference the edge, the id XML-Attribute is used.

It seems to me that the Multi(Di)Graph edge `key` value and any edge attributes in networkx are not strong enough to enforce uniqueness across the whole graph for the use case of Gephi.

Also, to preserve the original `key` value (edge uniqueness within multi-edges) in graphml, the first thought in my mind is 
to generate a new `id` (e.g. use a delimiter `id = "a_b_0"`, where `a` is the source, `b` is the target, and `0` is edge key) while writing edge information in graphml format and parse it later while reading the file, instead of only using edge `key` to represent `id` value. I am not sure whether it is a good idea. I am happy to discuss this issue further.
```python
import networkx as nx
G = nx.MultiDiGraph(((0, 1, 0), (0, 1, 1), (1, 2, 0), (2, 0, 0)))
nx.write_graphml(G, './test.graphml')
```
```xml
<?xml version='1.0' encoding='utf-8'?>
<graphml xmlns="http://graphml.graphdrawing.org/xmlns" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://graphml.graphdrawing.org/xmlns http://graphml.graphdrawing.org/xmlns/1.0/graphml.xsd">. 
  <graph edgedefault="directed"><node id="0"/>
    <node id="1"/>
    <node id="2"/>
    <edge id="0_1_0" source="0" target="1"/>
    <edge id="0_1_1" source="0" target="1"/>
    <edge id="1_2_0" source="1" target="2"/>
    <edge id="2_0_0" source="2" target="0"/>
  </graph>
</graphml>
```

**Comment by dschult** [comment]
I guess this rules out underscores in the node names.
We could try an id that mimics `"(node1, node2, key)"` which would work even if the nodes are tuples.
But I think the simpler route is to allow an edge attribute to be specified.

**Comment by berlincho** [comment]
Using the format, `"(node1, node2, key)"`, sounds like a good idea. 
I am thinking about adding a new input argument (e.g. edge_id) into `networkx.write_graphml()` and `networkx.generate_graphml()` to allow users to specify an edge attribute. If the argument is not specified, perhaps, we could set the `id` with `"(node1, node2, key)"` format.

My thought is like below
```python
import networkx as nx
G = nx.MultiDiGraph(((0, 1, 0, {'eid': 1}), (0, 1, 1, {'eid': 2}), (1, 2, 0, {'eid': 3}), (2, 0, 0, {'eid': 4})))
nx.write_graphml(G=G, path='./test.graphml', edge_id='eid')
```
Expected output
```xml
<?xml version='1.0' encoding='utf-8'?>
<graphml xmlns="http://graphml.graphdrawing.org/xmlns" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://graphml.graphdrawing.org/xmlns http://graphml.graphdrawing.org/xmlns/1.0/graphml.xsd">. 
  <key attr.name="eid" attr.type="long" for="edge" id="d0"/>
  <graph edgedefault="directed"><node id="0"/>
    <node id="1"/>
    <node id="2"/>
    <edge id="1" source="0" target="1">
      <data key="d0">1</data>
    </edge>
    <edge id="2" source="0" target="1">
      <data key="d0">2</data>
    </edge>
    <edge id="3" source="1" target="2">
      <data key="d0">3</data>
    </edge>
    <edge id="4" source="2" target="0">
      <data key="d0">4</data>
    </edge>
  </graph>
</graphml>
```
However, one issue I encountered is when I call `read_graphml()` later, the original `key` information (multi-edge uniquess) is missed.  
```python
G = nx.read_graphml('./test.graphml')
print(G.edges(keys=True, data=True))
#output [('0', '1', 1, {'eid': 1}), ('0', '1', 2, {'eid': 2}), ('1', '2', 3, {'eid': 3}), ('2', '0', 4, {'eid': 4})]
```

**Comment by berlincho** [comment]
This issue has been resolved in #4842 and should be closed.
